lib/pbench/server/auth/keycloak_admin.py

```diff
@@ -0,0 +1,331 @@
+                return self.get_user(user_id, self.user_realm)
+
+        self.add_header_param(key="Content-Type", value="application/json")
+        return self._post(
+            auth_urls.URL_ADMIN_USERS.format(**params_path), data=payload
+        ).json()
+
+    def delete_user(self, user_id: str) -> HTTPStatus:
+        """
+        Delete the given user entry
+        :param user_id: id of the user
+        :return:
+        """
+        params_path = {"realm-name": self.user_realm, "id": user_id}
+        return HTTPStatus(
+            self._delete(auth_urls.URL_ADMIN_USER.format(**params_path)).status_code
+        )
+
+    def update_user(self, user_id: str, payload: Dict) -> HTTPStatus:
+        """
+        Update the given user with payload data
+        https://www.keycloak.org/docs-api/18.0/rest-api/index.html#_userrepresentation
+        :param user_id: id of the user
+        :param payload: UserRepresentation
+        :return:
+        """
+        params_path = {"realm-name": self.realm_name, "id": user_id}
+        self.add_header_param(key="Content-Type", value="application/json")
+        return HTTPStatus(
+            self._put(
+                auth_urls.URL_ADMIN_USER.format(**params_path), data=payload
+            ).status_code
+        )
+
+    def realm_client_roles(self, realm: str = None, client_id: str = None) -> Dict:
+        """
+        Get all roles for the given realm and client.
+        If the client is not specified this will return all the roles under the realm
+        :param realm: realm id
+        :param client_id: id of the client, not the client name
+        https://www.keycloak.org/docs-api/18.0/rest-api/index.html#_rolerepresentation
+        :return: Keycloak server response (RoleRepresentation)
+        """
+        if not client_id:
+            params_path = {"realm-name": realm if realm else self.user_realm}
+            return self._get(
+                auth_urls.URL_ADMIN_REALM_ROLES.format(**params_path)
+            ).json()
+        else:
+            params_path = {
+                "realm-name": realm if realm else self.user_realm,
+                "client_id": client_id,
+            }
+            return self._get(
+                auth_urls.URL_ADMIN_CLIENT_ROLES.format(**params_path)
+            ).json()
+
+    def get_realm_client_role_members(
+        self,
+        role_name: str,
+        realm: str = None,
+        client_id: str = None,
+        params: Dict = None,
+    ):
+        """
+        Get all the members of the given role of the realm.
+        :param role_name: Name of the role.
+        :param realm: optional realm name
+        :param client_id: optional id of client
+        :param params: Additional Query parameters
+            (see https://www.keycloak.org/docs-api/18.0/rest-api/index.html#_roles_resource)
+        :return: Keycloak Server Response (UserRepresentation)
+        """
+        params = params or dict()
```

Comment:
Modifying the formal parameter is generally frowned upon; at best, it becomes non-obvious through most of the code what value to expect. Better to use a temporary.

(Also, technically, as written the type hint it `Optional[Dict[...]]`, not `Dict`. this is true for all of the parameters as they all are optional with a default of `None`...)